ops/model.py

```diff
@@ -606,6 +606,9 @@ def open_port(self, protocol: typing.Literal['tcp', 'udp', 'icmp'],
         However, normally charms should make the same open_port() call from
         every unit.
 
+        Use :func:`set_ports()` for a more declarative approach where all of
+        the ports that should be open are provided in a single call.
+
```

Comment:
Maybe this could also be communicated with [`typing.overload`](https://docs.python.org/3/library/typing.html#typing.overload).

Reply:
I'm pretty sure the `open-port` hook tool will give a decent error message -- but that might not map back to the Python arguments very well. So we might want to explicitly check and `raise TypeError` in `Unit.open_port` / `Unit.close_port`.

We could also add a `typing.overload` if it doesn't add too much complexity and if the docs still look sane.

Reply:
Ah, I wrote this before seeing your comment @benhoyt (TIL that "draft" has a bit of different meaning in Canonical than I'm used to :laughing:).

I first raised RuntimeError, and then noticed that ops.testing.Harness will raise ModelError, so I changed to that to be consistent. My inclination is that TypeError is best for `open_port('tcp')` but `open_port()` is more of a RuntimeError. However, I see that TypeError is raised in, for example, `get_secret` if the wrong argument combination is passed, so it would be consistent there. Overall, I would still generally prefer consistency with ops.testing.

> We could also add a typing.overload if it doesn't add too much complexity and if the docs still look sane.

I already did this too. It looks fairly sane I think:

![image](https://github.com/canonical/operator/assets/826522/10c68a06-fd82-4214-8501-33990b213b96)

Reply:
As discussed elsewhere in the PR, we decided to drop the change to have protocol default to 'tcp'. I've kept the checking that a port is not provided for icmp and is provided for tcp/udp, since that seems useful outside of the default.

@benhoyt do you think it's worth keeping the (simplified) overload? The docs look like this now:

![image](https://github.com/canonical/operator/assets/826522/a04ea26f-d4a1-4475-9c07-cfd296cb9204)

Basically, this means that if you do `open_port('icmp', 8000)` or `open_port('tcp')` your type checker will complain rather than it causing an error at run time. If you'd prefer to keep the docs simpler it's easy to drop those out as well.

Reply:
I don't mind this typing improvement. However:

1) Let's avoid repeating the docstring (it only needs to be on the last one -- see `Container.exec`). The body can be `...` for the others.
2) Let's just make sure we've tested that it type checks with both Pyright and Mypy. We probably need to enable type checking in our tests at some point, but for now, you can just test it manually.

Reply:
> 1. Let's avoid repeating the docstring (it only needs to be on the last one -- see `Container.exec`). The body can be `...` for the others.

When this is the case, the linting fails, because it's a public method without a docstring (it's what I originally did). Should I remove the docstring and add a pragma to tell the linter to ignore it?

> 2. Let's just make sure we've tested that it type checks with both Pyright and Mypy. We probably need to enable type checking in our tests at some point, but for now, you can just test it manually.

Sure. I've checked with pyright but will check mypy too.

Reply:
Ah yes, I guess that's what the `# noqa` comment is for in the `Container.exec` variants. That seems vastly preferable over repeating the docstring -- thanks.